src/sources/apache_metrics/parser.rs

```diff
@@ -0,0 +1,837 @@
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "could not parse value for {}: {}", self.key, self.err)
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(self.err.as_ref())
+    }
+}
+
+enum LineResult {
+    Metrics(Vec<Metric>),
+    Error(ParseError),
+    None,
+}
+
+fn line_to_metrics(
+    key: &str,
+    value: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> LineResult {
+    match key {
+        "ServerUptimeSeconds" => match value.parse::<f64>() {
+            Ok(value) => LineResult::Metrics(vec![Metric {
+                name: encode_namespace(namespace, "uptime_seconds_total"),
+                timestamp: Some(now),
+                tags: tags.map(|tags| tags.clone()),
+                kind: MetricKind::Absolute,
+                value: MetricValue::Counter { value },
+            }]),
+            Err(err) => LineResult::Error(ParseError {
+                key: key.to_string(),
+                err: err.into(),
+            }),
+        },
+        "Total Accesses" => match value.parse::<f64>() {
+            Ok(value) => LineResult::Metrics(vec![Metric {
+                name: encode_namespace(namespace, "access_total"),
+                timestamp: Some(now),
+                tags: tags.map(|tags| tags.clone()),
+                kind: MetricKind::Absolute,
+                value: MetricValue::Counter { value },
+            }]),
+            Err(err) => LineResult::Error(ParseError {
+                key: key.to_string(),
+                err: err.into(),
+            }),
+        },
+        "Total kBytes" => match value.parse::<u32>().map(|v| v * 1024) {
+            Ok(value) => LineResult::Metrics(vec![Metric {
+                name: encode_namespace(namespace, "sent_bytes_total"),
+                timestamp: Some(now),
+                tags: tags.map(|tags| tags.clone()),
+                kind: MetricKind::Absolute,
+                value: MetricValue::Counter {
+                    value: value.into(),
+                },
+            }]),
+            Err(err) => LineResult::Error(ParseError {
+                key: key.to_string(),
+                err: err.into(),
+            }),
+        },
+        "Total Duration" => match value.parse::<f64>() {
+            Ok(value) => LineResult::Metrics(vec![Metric {
+                name: encode_namespace(namespace, "duration_seconds_total"),
+                timestamp: Some(now),
+                tags: tags.map(|tags| tags.clone()),
+                kind: MetricKind::Absolute,
+                value: MetricValue::Counter { value }, // TODO verify unit
+            }]),
+            Err(err) => LineResult::Error(ParseError {
+                key: key.to_string(),
+                err: err.into(),
+            }),
+        },
+        "CPUUser" => match value.parse::<f64>() {
```

Comment:
I see a reasonable amount of duplication in the next couple of match arms.

Perhaps use a single arm, and then split up where needed? e.g.

```rust
"CPUUser " | "CPUSystem" | /* ... */ => {
	// ...

	match { /* ... */ }
}
```

Can't say for certain how much clean-up that would provide, but might be worth a try.

Reply:
It is a bit duplicative, but I actually like the simplicity of just mapping each string rather than parsing the key names. If you feel strongly though, I'll give it a shot.